### ISSUE TITLE
IPv6 implementation

### ISSUE THREAD

**Issue by alberk8** [hamishforbes/lua-resty-iputils#4]
Hi,
Is there any progress in the IPv6 implementation?  Thank you.

**Comment by hamishforbes** [comment]
Hi,
 Unfortunately not, I don't have a need for an ipv6 version yet myself so its hard to prioritise that work.
Pull requests are most welcome though!

**Comment by GUI** [comment]
If it helps, here's a slightly different Lua library I put together that supports IPv6 and CIDR comparisons (as well as IPv4): https://github.com/GUI/lua-libcidr-ffi The downside is that it requires [libcidr](http://www.over-yonder.net/%7Efullermd/projects/libcidr) to be installed on the system, and the API is probably a bit more raw that what resty-iputils provides.

I had originally looked to add IPv6 compatibility to resty-iputils directly, but I couldn't find a good way to efficiently compare IPv6 CIDR ranges in pure Lua (I found a couple different implementations in other open source projects, but their performance wasn't ideal). So I ended up writing the small FFI wrapper around libcidr (which is what does the real work). It would be possible to leverage the libcidr-ffi library in resty-iputils for IPv6 compatibility, but that would mean making libcidr a dependency, and I'm not sure if that's desirable.

**Comment by hamishforbes** [comment]
Cool!
Yeah I think i'll leave this lib as a pure Lua version, if people don't mind the dependency they can use your libcidr based one.

If you don't mind I'll link to your repo on the README here?

**Comment by GUI** [comment]
:+1: Sounds good. And feel free to add a link--thanks!

**Comment by GUI** [comment]
Good deal. I've also added a link back to lua-resty-iputils from lua-libcidr-ffi's README. Hopefully that will help people find the library that best fits their needs regardless of which one they stumble into first.

**Comment by hamishforbes** [comment]
:+1: cool, IPv6 is on the roadmap for the dayjob this year so maybe i'll find time to add a pure Lua version then :)

**Comment by joshenders** [comment]
Hi @hamishforbes ever find time to add this?

**Comment by hamishforbes** [comment]
No sorry, not yet.

**Comment by joshenders** [comment]
I might me able to pick this up. Is there benefit in a pure lua implementation or would it make more sense to extend lua-libcidr-ffi to be more user-friendly?

**Comment by hamishforbes** [comment]
The main benefit is fewer dependencies, you don't need libcidr installed on the system for the pure Lua version.

There might be performance differences, one way or the other, but I haven't benchmarked them.

**Comment by GUI** [comment]
Yeah, libcidr also isn't a particularly common system dependency (I don't think there's Ubuntu or CentOS packages for it), so while I've found lua-libcidr-ffi useful for systems I can manually install libcidr on, I also agree that something with fewer dependencies is nice.

If it helps anyone looking at pure Lua options, here's some more specific details about the alternatives I found before building lua-libcidr-ffi:

- I found a couple IP libraries with IPv6 support in pure Lua that are embedded inside larger projects:
  - https://github.com/openwrt/luci/blob/0.11.1/libs/core/luasrc/ip.lua
  - https://github.com/nmap/nmap/blob/03f823d6c53a18a10e47f527019548dd2412dd32/nselib/ipOps.lua
- The library from "luci" is actually from an older version. They've since replaced it with a C library with Lua bindings (which I think might be more standalone versus needing a system library like libcidr): https://github.com/openwrt/luci/blob/ec9942925c80a278a565091a194485999d6be74d/libs/luci-lib-ip/src/ip.c https://github.com/openwrt/luci/blob/ec9942925c80a278a565091a194485999d6be74d/libs/luci-lib-ip/src/ip.luadoc
- When I benchmarked the pure Lua implementations against libcidr-ffi, libcidr-ffi was quite a bit faster, particularly for IPv6 (sorry I don't have any numbers handy anymore). But I'll put a big caveat around this, since I was just hacking around with copying and pasting those existing Lua implementations and trying to make them work--I'm not sure if what I managed to implement was very optimized. A more optimized pure Lua approach might be possible. And while the performance difference in micro-benchmarks was noticeable, I'm not sure if it would really have that much of a real world impact.

**Comment by joshenders** [comment]
That's funny, I was looking at luci as well. We have a need for IPv6 support in our environment at Pinterest. If we take this on, it'll probably be in Q2. I'll make sure we contribute this back; seems like this is sorely needed.

**Comment by bungle** [comment]
I recently forked [mediador](https://github.com/simbiose/mediador) library here:
https://github.com/mashape/lua-resty-mediador

It does support IPv6 and IPv4. Haven't benchmarked it yet.
